Add abitlity to resize peerTable and banTable

shaavan opened the pull request:
Till now, peerTable and banTable under the peerTab were fixed in height and cannot be resized. This creates unnecessary inconvenience, especially when you have many items on one table and not many on another.

This PR proposes to add a splitter between these two tables, allowing them to be resized freely.

|Master|PR|
|---------|----|
|![Screenshot from 2021-10-25 21-26-09](https://user-images.githubusercontent.com/85434418/138748172-020daf23-a07b-48aa-8df6-22666892cf38.png)|![Screenshot from 2021-11-01 17-43-01](https://user-images.githubusercontent.com/85434418/139672087-302d7cc1-3400-46d0-b258-df24f1266d3c.png)|

shaavan commented on src/qt/forms/debugwindow.ui:
```diff
@@ -887,85 +887,113 @@
+               <item>
+                <widget class="QLabel" name="peerHeading_2">
+                 <property name="text">
+                  <string>Connected peers</string>
```
> This shouldn't display until there is a Banned Peers table
Additionally, I shouldn't be able to hide any of the tables

These are valid suggestions. Let me take a look into implementing these.

>I don't think this should be a part of this PR considering the title suggests you want to just add functionality to resize tabled

Ok then, I shall remove the name change from this PR and may open it as another pull request in the future (if need be)

shaavan commented:
Updated 5e54cfe to c09e6ad ([pr457.01](https://github.com/shaavan/gui/tree/pr457.01) -> [pr457.02](https://github.com/shaavan/gui/tree/pr457.02))
Addressed @jarolrod comments

Updates:
1. Considering that naming peerTable was not gathering a lot of support. I have removed that change from this PR.
2. Made the peerTable and banTable non-collapsible.
3. Increased splitter’s height from `3px` to `4px`.

> This shouldn't display until there is a Banned Peers table

@jarolrod. As you rightly mentioned, showing splitter when no BanTable was looking kind of off. But in the new iteration of this PR, the splitter seems like an element of peerTable and looks aesthetically pleasing even when there is no banTable. I have updated the PR description with the new Screenshots. You can look at them to better get the feel of what I am trying to say.

jarolrod reviewed (commented):
Getting there, but the splitter should behave like the splitter that currently exists between the `Peers Table` and the `Peer Information Pane`.

When I select a Peer, I get the `Peer Information Pane`, as well as a splitter where I can resize as needed. When the peer is de-selected, the splitter disappears; because there is nothing to resize.

Similarly, until I have a banned peer and a banned peer table, the splitter being introduced here should not be displayed at all because there is nothing to resize. Additionally, if i unban a peer and subsequently no longer have a banned peers table, the splitter should disappear as well.

shaavan commented:
Updated from c09e6ad39da24a7a969533f6ce5c9cea14e8c9ce to aa6229e5d2bebe20d3e8e3eded8ff3ad7841bfcb ([pr457.02](https://github.com/shaavan/gui/tree/pr457.02) -> [pr457.03](https://github.com/shaavan/gui/tree/pr457.03), [diff](https://github.com/shaavan/gui/compare/pr457.02..pr457.03))
Addressed @jarolrod comments

**Changes:**

- Instead of hiding individual children of banTable. The code is now updated to hide the banTable widget. This will make sure the splitter will not be displayed when there is no banTable.
- Updated sizePolicy to better reflect the peerWidget and the banTable widget.
- Added code to save and restore the state (position) of the splitter. The state is remembered even in the case when GUI is shut down and restarted.

shaavan commented:
cc @jarolrod. PR ready for another review!

kristapsk commented on src/qt/forms/debugwindow.ui:
```diff
@@ -6,13 +6,16 @@
     <y>0</y>
-    <width>740</width>
-    <height>430</height>
+    <width>744</width>
+    <height>531</height>
```
Why this change?

shaavan replied:
I am not sure of the reason. But I observed they are set to these values (744, 531) each time debugwindow.ui is opened in QTCreator even if we try to change them. However, this does not cause any functional change.

kristapsk replied:
Ok, then seems reasonable to change to these values.

RandyMcMillan replied:
We manually edit these values to remove this diff junk - this way it stays consistent.

RandyMcMillan replied:
Including - resetting the currentIndex to zero (below)

shaavan replied:
Resetted these changes to the master's value in 0c364201a2e16355cd6424c6137761c7967db4e0

kristapsk reviewed (approved):
ACK aa6229e5d2bebe20d3e8e3eded8ff3ad7841bfcb (tested under Linux / Xfce4)

promag commented on src/qt/forms/debugwindow.ui:
```diff
@@ -887,85 +890,112 @@
-              <set>Qt::NoTextInteraction</set>
+            <widget class="QSplitter" name="tableSplitter">
+             <property name="styleSheet">
+              <string notr="true">QSplitter::handle{background: #a6a6a6;}</string>
```
Why this style?

shaavan replied:
I did this to make the splitter visible and looks aesthetically pleasing.
Do you recommend keeping the splitter invisible?

shaavan replied:
What do you think about this? @promag

hebasto replied:
> Do you recommend keeping the splitter invisible?

I don't think the hardcoded color will work for any platform and appearance (dark/light), no?

shaavan replied:
> I don't think the hardcoded color will work for any platform and appearance (dark/light), no?

The spiller looks fine for the default light mode (in Linux) and dark mode in macOS, which can be seen in the screenshots of the following comments: [light](https://github.com/bitcoin-core/gui/pull/457#issue-1035432748); [dark](https://github.com/bitcoin-core/gui/pull/457#pullrequestreview-791298779).

However, this doesn't guarantee that this will work for any custom appearance. Do you recommend removing the hardcoded color and making the splitter invisible again?

hebasto replied:
Why not keeping the same look as for horizontal `splitter`?

> Do you recommend keeping the splitter invisible?

What does make you think so?

shaavan replied:
> Do you recommend keeping the splitter invisible?

By this, I meant to say, keeping it the same as the horizontal splitter.

> Why not keep the same look as for horizontal splitter?

I think this will be a more optimal approach to go with.

hebasto replied:
But the horizontal splitter is not invisible :)

And having different looks for similar GUI elements does not as an optimal design for me.

hebasto replied:
At least, introducing a new GUI element, and changing thuis GUI element style across the GUI are two different tasks.

shaavan replied:
Fixed splitter color to default palette in 0c364201a2e16355cd6424c6137761c7967db4e0

promag reviewed (commented):
Concept ACK.

w0xlt reviewed (approved):
tACK aa6229e . Tested on Ubuntu 21.10.

hebasto commented:
Concept ACK.

hebasto commented on src/qt/rpcconsole.cpp:
```diff
@@ -750,6 +754,9 @@ void RPCConsole::setClientModel(ClientModel *model, int bestblock_height, int64_
         ui->startupTime->setText(model->formatClientStartupTime());
         ui->networkName->setText(QString::fromStdString(Params().NetworkIDString()));
 
+        //Setup Peer and Ban Table Splitter
+        ui->tableSplitter->setChildrenCollapsible(false);
+
```
This code could be a part of the `debugwindow.ui` file as it is for the horizontal `splitter`.

Qt Designer tool could very helpful for such tasks.

shaavan replied:
Addressed in 0c364201a2e16355cd6424c6137761c7967db4e0

hebasto commented on src/qt/forms/debugwindow.ui:
```diff
@@ -36,7 +39,7 @@
    <item>
     <widget class="QTabWidget" name="tabWidget">
      <property name="currentIndex">
-      <number>0</number>
+      <number>3</number>
```
Why this change?

shaavan replied:
This was an auto-generated change by QtCreator. Fixed in 0c364201a2e16355cd6424c6137761c7967db4e0

shaavan commented:
Updated from aa6229e5d2bebe20d3e8e3eded8ff3ad7841bfcb to 0c364201a2e16355cd6424c6137761c7967db4e0([pr457.04](https://github.com/shaavan/gui/tree/pr457.04) -> [pr457.05](https://github.com/shaavan/gui/tree/pr457.05))
Addressed @hebasto and @randymacmilan comments

- Removed custom color of the added splitter
- Moved `childrenCollapsible=False` to debugwindow.ui file.
- Removed auto-generated changes by QtCreator.

hebasto commented on src/qt/forms/debugwindow.ui:
```diff
@@ -887,85 +887,112 @@
            </size>
           </property>
-          <layout class="QVBoxLayout" name="verticalLayout_7">
+          <layout class="QVBoxLayout" name="verticalLayout_9">
```
Is this change required?

shaavan replied:
Since this PR introduced a new sub-Layout, the introduced sub-Layout was named in the chronological order of its introduction. And since **verticalLayout_7** already exists (on Line 935), giving this the same name leads to a warning message while compiling:

```
qt/forms/debugwindow.ui: Warning: The name 'verticalLayout_7' (QVBoxLayout) is already in use, defaulting to 'verticalLayout_71'.
```

So, I am letting this change be as it is for now, and I welcome any suggestions you might have.

hebasto reviewed (commented):
Approach ACK 0c364201a2e16355cd6424c6137761c7967db4e0.

Could you please separate a refactoring commit which introduces the `banTable` widget?

shaavan commented:
Updated from 0c364201a2e16355cd6424c6137761c7967db4e0 to 5a9bdce0ad753919522b8bc627c1c12d92d7cae8 ([pr457.05](https://github.com/shaavan/gui/commits/pr457.05) -> [pr457.06](https://github.com/shaavan/gui/commits/pr457.06), [diff](https://github.com/shaavan/gui/compare/pr457.05..pr457.06))

Addressed @hebasto comment

Changes:
- Rebased PR over the current master.
- Separated adding of **banTable** and **tableSplitter** in separate commits.
- Removed unnecessary `width`/`height` changes.

jarolrod reviewed (commented):
While the resizing works, the UI here does not. The UI element representing the splitter is too small and is pinned to the bottom of `peerTable`.

| linux | macOS |
| ----- | ------- |
![Screenshot from 2022-08-06 18-41-26](https://user-images.githubusercontent.com/23396902/183268373-a6aa5891-568d-4df9-a56e-ded4419c4361.png) | <img width="965" alt="Screen Shot 2022-08-06 at 6 49 31 PM" src="https://user-images.githubusercontent.com/23396902/183268379-383c1bc6-394a-453f-abdd-681f819418b8.png"> |

hebasto commented:
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.
